pkg/controller/installation/windows.go

```diff
@@ -0,0 +1,379 @@
+	calicoWindowsUpgradingTaint = &v1.Taint{
+		Key:    common.CalicoWindowsUpgradeTaintKey,
+		Effect: v1.TaintEffectNoSchedule,
+	}
+)
+
+// calicoWindowsUpgrader helps manage the upgrade of Calico Windows nodes.
+// It works in conjunction with the CalicoUpdate service running on each node.
+type calicoWindowsUpgrader struct {
+	clientset            kubernetes.Interface
+	client               client.Client
+	statusManager        status.StatusManager
+	requestChan          chan<- utils.ReconcileRequest
+	nodeInformer         cache.Controller
+	nodeIndexer          cache.Indexer
+	nodesToUpgrade       []*corev1.Node
+	nodesUpgrading       []*corev1.Node
+	nodesFinishedUpgrade []*corev1.Node
+}
+
+func (w *calicoWindowsUpgrader) hasPendingUpgrades() bool {
+	return len(w.nodesToUpgrade)+len(w.nodesUpgrading) > 0
+}
+
+func (w *calicoWindowsUpgrader) triggerReconcileRequest(obj interface{}) {
+	node := obj.(*corev1.Node)
+
+	// Ignore non-windows nodes or nodes without an os label.
+	if os, ok := node.GetLabels()[corev1.LabelOSStable]; !ok || os != "windows" {
+		return
+	}
+
+	req := utils.ReconcileRequest{
+		Context:    context.Background(),
+		Request:    reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}},
+		ResultChan: make(chan utils.ReconcileResult),
+	}
+
+	w.requestChan <- req
+}
+
+// newCalicoWindowsUpgrader creates a Calico Windows upgrader.
+func newCalicoWindowsUpgrader(cs kubernetes.Interface, c client.Client, nodeListWatch cache.ListerWatcher, statusManager status.StatusManager, requestChan chan<- utils.ReconcileRequest) *calicoWindowsUpgrader {
+	w := &calicoWindowsUpgrader{
+		clientset:     cs,
+		client:        c,
+		statusManager: statusManager,
+		requestChan:   requestChan,
+	}
+
+	handlers := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			w.triggerReconcileRequest(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			w.triggerReconcileRequest(newObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			w.triggerReconcileRequest(obj)
+		},
+	}
+	w.nodeIndexer, w.nodeInformer = cache.NewIndexerInformer(nodeListWatch, &corev1.Node{}, 0, handlers, cache.Indexers{})
+
+	return w
+}
+
+func (w *calicoWindowsUpgrader) upgradeWindowsNodes(expectedProduct operatorv1.ProductVariant) error {
+	expectedVersion := common.WindowsLatestVersionString(expectedProduct)
+	windowsLog.V(1).Info(fmt.Sprintf("Expected version: %v", expectedVersion))
+
+	// For now, do nothing if the upgrade version is not the upcoming
+	// Enterprise v3.11.0. This prevents a user from accidentally triggering an
+	// in-place upgrade from Calico v3.21.0 to Enterprise v3.10.x. That upgrade
+	// would fail since the calico windows upgrade is only supported in
+	// Enterprise v3.11.x and newer.
+	// TODO: remove this once Enterprise v3.11.0 is released.
+	if expectedProduct == operatorv1.TigeraSecureEnterprise && !strings.HasPrefix(expectedVersion, "Enterprise-v3.11.0") {
```

Comment:
I still need to update this as discussed... we want to avoid doing an explicit version check here